### PR TITLE
Defeating the windmill

### PR DESCRIPTION
**Pull request by soujiro32167**
This is a first pass at decoupling (and rewriting) the jax-rs web service in challenge-1.

The premise for scaling here is two fold: 

1. horizontal/vertical scaling of databases using AWS databases (mysql in this instance, but applies to any managed db)
2. horizontal/vertical scaling of the application engine using a serverless framework - AWS Lambda and the serverless deployment framework (https://serverless.com)

My initial attempt was to keep the application as-is in Java, hooking it up to the Serverless event trigger. However, after running into a few blocks (Lambda does not read .war packages, .jar packaging failing, madness descending), I've decided to jump ship and rewrite the thing in Node.

## Improvements needed

1. Load test to prove scaling: so far, I've run 10, 100, 1000 concurrent postman requests on the endpoint. Average response time is stable around 65ms in all cases. This, of course, needs to be scaled up and made reproducible
2. Automatic DB provision: Creating the database and table(s) using the serverless.yml file. This was in the Dynamo DB tutorial but somehow for RDS it not as simple...
3. Credentials management for DB: no one likes unencrypted credentials lying around in source control...
4. Containerization: Another way of scaling this would be inside a Docker container as part of a Kubernetes or Marathon cluster. This is something I'm hoping to try when I get back end of March
5. Identity management: hooking the service up to Cognito or a custom identity management layer
6. CI/CD: versioning artifacts, versioning the API, build server and so on

In conclusion, please consider this an initial effort, before I head off into the jungles of Siem Reap.

Happy reviewing!

Eli

### CODE REVIEW

**Comment by bruno-morel** [comment]
Good start....can we see more ? :)
I also believe some part of this code are exposed to an SQL injection attach...maybe prove me wrong or fix it ? :)

**Comment by soujiro32167** [comment]
Tested for SQL injection:

- `GET /reminders?name=;DROP TABLE reminders`
- `GET /reminders?name=asd OR 1=1`
- ```
POST /reminders/213
{
  "name": "; DROP TABLE reminders",
  "--; DROP TABLE reminders": "asd"
}
```

They are not successful. This is because all of the queries use `mysql.escape` and `mysql.escapeId` either explicitly (when composing from request parameters) or implicitly (when passing values to replace '?')

**Comment by soujiro32167** [comment]
Also, I'm having a fascinating read about about db connection management and serverless design. Since there is no state, there is no (simple, obvious) way to reuse an open connection. In AWS specifically, the VM that runs the lambda does in fact stay alive for 300 seconds [citation needed], so using a connection pool is not a bad idea. Anecdotally, with 1000 concurrent requests, # of connections did not exceed ~8. So much learning :)

**Comment by soujiro32167** [comment]
First load test complete: 20 requests per second for 60 seconds
```
All virtual users finished
Summary report @ 13:59:08(-0400) 2018-03-26
  Scenarios launched:  1200
  Scenarios completed: 1200
  Requests completed:  1200
  RPS sent: 19.69
  Request latency:
    min: 80.2
    max: 11081.3
    median: 98
    p95: 147.7
    p99: 1416.7
  Scenario counts:
    0: 1200 (100%)
  Codes:
    200: 1200
```

**Comment by soujiro32167** [comment]
Looks like I spoke too soon. A test with 15000 users has hit the DB connection limit:

`Error: ER_CON_COUNT_ERROR: Too many connections`

Looks like I'll have to open and close the connection each query

**Comment by soujiro32167** [comment]
Turns out my machine couldn't open 15,000 connections on its own anyways....but the connection issue is now fixed. This is the highest RPS I've tested so far, and the results are not bad for free-tier DB and Lambda, without any caching:

Report @ 18:33:33(-0400) 2018-03-26
  Scenarios launched:  811
  Scenarios completed: 791
  Requests completed:  792
  RPS sent: 81.1
  Request latency:
    min: 141.2
    max: 1368.4
    median: 222.4
    p95: 358.1
    p99: 497.4
  Codes:
    200: 388
    201: 398
    502: 6

**Comment by bruno-morel** [comment]
We got good news... poke the Galyna :)

**Comment by soujiro32167** [comment]
Initiating pokage! Thanks for the heads up
